src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java

```diff
@@ -416,6 +421,32 @@ private KNNEngine validateDimensions(final KNNMethodContext knnMethodContext, fi
             }
             return knnEngine;
         }
+
+        /**
+         * Validate whether provided full field name contain any invalid characters for physical file name.
+         * At the moment, we use a field name as a part of file name while we throw an exception
+         * if a physical file name contains any invalid characters when creating snapshot.
+         * To prevent from this happening, we restrict vector field name and make sure generated file to have a valid name.
+         *
+         * @param context : Builder context to have field name info.
+         */
+        private void validateFullFieldName(final BuilderContext context) {
```

Comment:
is this check happening during every document ingestion or happening once during index creation?

Reply:
That depends on the whether dynamic mapping is enabled.
If it was enabled, then `MapperService` will walk through the documents to identify whether there are any new fields added, and during which, the check will be done along with other validation checks.
If it was not enabled, it won't be called.

Reply:
@0ctopus13prime make sense. What tests we have performed to validate this behavior? Reason I am asking this is if this check is performed per document then this check will incur latency in the overall indexing time. Hence I want to make sure that this check is done once.

Reply:
I don't think it will add up trivial latency because of the changes, as the validation check that I added is merely walk through field char[] to look up hash set. Overall time complexity is bounded by `O(N)` where N = the number of chars in vector field. Presumably this will only add up nano scale latency in the end, but only affecting vector fields.

But if we're being conservative, then I can run TPS benchmark and compare the metrics between baseline.
What's your thought?

Reply:
Yeah I understand the latency is in nanos. But would be interesting is if you can validate/provide details how you tested to know that this function runs once. We don't need benchmarks here.

Reply:
Mostly done in my local dev environment with debugger attached in IDE.

Launch a server in IDE (I got rid of all segments and repos every time I launch a new server)
1. Tried to create a vector index with invalid vector field name. Ex: `my vector`. And validated the request is failing.
```
curl -X PUT 'http://localhost:9200/knn-index/' -H 'Content-Type: application/json' -d '
{
  "settings": {
    "index": {
      "knn": true,
      "knn.algo_param.ef_search": 100
    }
  },
  "mappings": {
    "properties": {
      "my vector": { <------------ This must fail
        "type": "knn_vector",
        "dimension": 2
      }
    },
    "dynamic": false | true <-------- Enable or disable dynamic mapping
  }
}
'
```

Response:
```
{
  "error": {
    "root_cause": [
      {
        "type": "resource_already_exists_exception",
        "reason": "index [knn-index/kOKM8elaSeacwnNGF11ZMg] already exists",
        "index": "knn-index",
        "index_uuid": "kOKM8elaSeacwnNGF11ZMg"
      }
    ],
    "type": "resource_already_exists_exception",
    "reason": "index [knn-index/kOKM8elaSeacwnNGF11ZMg] already exists",
    "index": "knn-index",
    "index_uuid": "kOKM8elaSeacwnNGF11ZMg"
  },
  "status": 400
}
```

2. Tried to ingest with both cases where dynamic mapping is enabled. In this case, I created a vector index with field name `my_vector` (valid one)
```
curl -X POST 'http://localhost:9200/_bulk' -H 'Content-Type: application/json' -d '{ "index": { "_index": "knn-index", "_id": "1" } }
   { "my_vector": [1.5, 2.5], "price": 12.2 }
   { "index": { "_index": "knn-index", "_id": "2" } }
   { "my_vector": [2.5, 3.5], "price": 7.1 }
   { "index": { "_index": "knn-index", "_id": "3" } }
   { "my_vector": [3.5, 4.5], "price": 12.9 }
   { "index": { "_index": "knn-index", "_id": "4" } }
   { "my_vector": [5.5, 6.5], "price": 1.2 }
   { "index": { "_index": "knn-index", "_id": "5" } }
   { "my_vector": [4.5, 5.5], "price": 3.7 }
'
```

Validated : 
- `validateFullFieldName` was called for each `my_vector`.


I tried to put everything I have tested so far, but if you need additional info, please feel free to drop a comment!

Reply:
@0ctopus13prime this tells me that for every document indexing this function will be called. What would be the case in which this function won't be called.

Reply:
@navneet1v 
As I said, if `dynamic` is disabled, the validation check that I add will not be called.
But if `dynamic` is enabled, during the document parsing, it will invoke the function along with others.

```
        public KNNVectorFieldMapper build(Mapper.BuilderContext context) {
            this.validateFullFieldName(context); <------- This
            KNNMethodContext knnMethodContext = (KNNMethodContext)this.knnMethodContext.getValue();
            this.setDefaultSpaceType(knnMethodContext, (VectorDataType)this.vectorDataType.getValue());
            this.validateSpaceType(knnMethodContext, (VectorDataType)this.vectorDataType.getValue());
            this.validateDimensions(knnMethodContext, (VectorDataType)this.vectorDataType.getValue());
            this.validateEncoder(knnMethodContext, (VectorDataType)this.vectorDataType.getValue());
```